German way property set and improvements to bicycle routes

leonardehrenfried opened the pull request:
### Summary
This PR improves the cycling algorithm by making the following improvements:

- it applies the tags `lcn`, `rcn`, `ncn`  from OSM cycling network relations to the relation's members
- since ways can be on all of these networks at the same time a logical OR matcher was introduced for the way property sets
- a German way property set was introduced that fine-tunes the routing for Germany but also takes advantage of the above code
- a more realistic turn cost model for bicycles that avoids left turns
- makes the car speed parsing more robust and allow single digit speeds

### Issue
closes #3397 
closes #3398
closes #3402

### Unit tests

There are a tests covering the way property set and the logical OR.

I've also added a new test for applying the bike route relation tags to the ways.

### Code style

✅

### Documentation

I've added the way property set to the documentation and updated the documentation about how to prepare OSM data to include bicycle routes.

### Changelog

✅

optionsome commented:
Haven't taken a deep look at this but is "a more realistic turn cost model for bicycles that avoids left turns" still a valid concern in countries that use left-hand traffic?

leonardehrenfried commented:
This exact question came up in the dev meeting and I will make it respect left hand traffic.

leonardehrenfried commented:
I've taken the review feedback from todays meeting on board and changed the following things:

- Left-hand/right-hand side traffic is now taken into account. This meant adding a property to the graph which is used to initialise the routing request.
- @abyrd request of moving the relation check into the method.

Now that there is a another field in the graph (driving direction), do we need to increment the serialization version?

flaktack commented on src/main/java/org/opentripplanner/routing/core/SimpleIntersectionTraversalCostModel.java:
```diff
@@ -101,12 +108,25 @@ private double computeCyclingTraversalCost(IntersectionVertex v, StreetEdge from
         var turnAngle = calculateTurnAngle(from, to, request);
```
`RoutingRequest` also contains a field for the driving direction ([`driveOnRight`](https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/src/main/java/org/opentripplanner/routing/api/request/RoutingRequest.java#L589-L593)), which is used by `calculateTurnAngle()`. The code here seems to work independently of that.

leonardehrenfried replied:
Well spotted! 

Am I reading it correctly that when you set the driving direction then to left hand side, then a left turn is magically turned into a right hand one? If this is the case then I don't need change anything.

leonardehrenfried replied:
Ok, after a bit of investigation, this does exactly what we need. I've added tests to verify.

leonardehrenfried commented:
It turns out that left/hand right hand traffic was already taken into account however the method and variable names made it hard to find out. I've added tests to make sure that it really is the case.

I've also incorporated @abyrd's feedback.

leonardehrenfried commented:
@abyrd I've removed the the request parameter `driveOnRight` and replaced it with an enum derived from the way property sets. I've also refactored the code that deals with left/right turns, removed the magic and made it generally more clear what is happening.

I've caught more cases where someone uses logical ORs with an unsupported combination and an exception is thrown.

@t2gran @abyrd @optionsome @gmellemstrand Since the improvements of the bicycle routing rests mainly on adding the mixin for lcn/rcn/ncn way, should I add this to the default way property set?

t2gran commented:
> Since the improvements of the bicycle routing rests mainly on adding the mixin for lcn/rcn/ncn way, should I add this to the default way property set? 

:+1: From me

t2gran commented on src/main/java/org/opentripplanner/graph_builder/module/osm/GermanyWayPropertySetSource.java:
```diff
@@ -0,0 +1,119 @@
+
+    @Override
+    public void populateProperties(WayPropertySet props) {
+        // Replace existing matching properties as the logic is that the first statement registered takes precedence over later statements
+
+        /*
+         * Automobile speeds in Germany. General speed limit is 50kph in settlements, 100kph outside settlements.
+         * For motorways, there (currently still) is no limit. Nevertheless 120kph is assumed to reflect varying
+         * traffic conditions
+         *
+         */
```
Code style: 

4.8.6.1 On multi-line `/* ... */` comments we do not begin the intermediate lines with asterisks `*`.

I personally avoid using C-style comments mixed with code, the only place I use it is to make sections in a class like /* private methods */.

t2gran commented on src/main/java/org/opentripplanner/graph_builder/module/osm/GermanyWayPropertySetSource.java:
```diff
@@ -0,0 +1,119 @@
+        props.setProperties("maxspeed=100", StreetTraversalPermission.ALL, 5.0, 5.0, true);
+
+        /* tracktype */
+        props.setProperties(
+                "tracktype=grade1", StreetTraversalPermission.ALL, 1.0, 1.0, true); // Solid
+        props.setProperties(
+                "tracktype=grade2", StreetTraversalPermission.ALL, 1.1, 1.1,
+                true
+        ); // Solid but unpaved.
+        props.setProperties(
```
Line comments should be before the statment, never use line comments in the right margin. Above is a good example why. It was not obvious to me first if the `// Solid but unpaved.` was commenting on the line before or after.

leonardehrenfried commented:
I've just taken on the feedback from today's meeting and the formatting feedback by @t2gran.

I think this is good for another round of reviews.

leonardehrenfried commented:
Unless there are more review comments, I think this one is ready to merge, isn't it?
